2023/FerrariStanganini/NetManProject/src/main.c

```diff
@@ -0,0 +1,547 @@
+{
+    DATA *data = (DATA *)d;
+    free(key);
+    roaring_bitmap_free(data->bitmap);
+    free(data);
+}
+
+/*************************************************/
+
+void my_sigalarm(int sig)
+{
+    print_stats();
+    alarm(ALARM_SLEEP);
+    signal(SIGALRM, my_sigalarm);
+}
+
+/*************************************************/
+
+char *proto2str(u_short proto)
+{
+    static char protoName[8];
+    switch (proto)
+    {
+        case IPPROTO_TCP:
+            return ("TCP");
+        case IPPROTO_UDP:
+            return ("UDP");
+        case IPPROTO_ICMP:
+            return ("ICMP");
+        default:
+            snprintf(protoName, sizeof(protoName), "%d", proto);
+            return (protoName);
+    }
+}
+
+/*************************************************/
+
+void getBroadCast(char *device)
+{
+    struct ifaddrs *ifaddr, *ifa;
+    struct sockaddr_in *sa = NULL;
+    struct sockaddr_in *su = NULL;
+    char subnet_mask[INET_ADDRSTRLEN];
+    char ip[INET_ADDRSTRLEN];
+    char broad[INET_ADDRSTRLEN];
+
+    if (getifaddrs(&ifaddr) == -1)
+    {
+        perror("getifaddrs");
+        exit(EXIT_FAILURE);
+    }
+    for (ifa = ifaddr; ifa != NULL; ifa = ifa->ifa_next)
+    {
+        if (strcmp(ifa->ifa_name, device) == 0 && ifa->ifa_addr->sa_family == AF_INET)
+        {
+            sa = (struct sockaddr_in *)ifa->ifa_netmask;
+            su = (struct sockaddr_in *)ifa->ifa_addr;
+            inet_ntop(AF_INET, &(sa->sin_addr), subnet_mask, INET_ADDRSTRLEN);
+            inet_ntop(AF_INET, &(su->sin_addr), ip, INET_ADDRSTRLEN);
+            break;
+        }
+    }
+
+    inet_pton(AF_INET, ip, &myIP.sin_addr);
+    broadcastIP.sin_addr.s_addr = su->sin_addr.s_addr | ~(sa->sin_addr.s_addr);
+    inet_ntop(AF_INET, &(broadcastIP.sin_addr), broad, INET_ADDRSTRLEN);
+    printf("\nSubnetMask:%s \nLocalIP:%s \nBroadcastIP: %s\n", subnet_mask, ip, broad);
+    freeifaddrs(ifaddr);
+}
+
+/*************************************************/
+
+void dummyProcesssPacket(u_char *_deviceId, const struct pcap_pkthdr *h, const u_char *p)
+{
+    struct ether_header ehdr;
+    struct ip ip;
+    struct tcphdr tcp;
+
+    memcpy(&ehdr, p, sizeof(struct ether_header));
+    memcpy(&ip, p + sizeof(ehdr), sizeof(struct ip));
+    u_short eth_type = ntohs(ehdr.ether_type);
+
+    if (eth_type == 0x0800)
+    {
+        if (ip.ip_p == IPPROTO_TCP &&
+                (ntohl(ip.ip_dst.s_addr) != ntohl(broadcastIP.sin_addr.s_addr)) &&
+                (ntohl(ip.ip_dst.s_addr) != ntohl(intraIP.sin_addr.s_addr)) &&
+                //(ntohl(ip.ip_dst.s_addr) != ntohl(myIP.sin_addr.s_addr)) &&
+                //(ntohl(ip.ip_src.s_addr) != ntohl(myIP.sin_addr.s_addr)) &&
+                (ntohl(ip.ip_dst.s_addr) != ntohl(allbroadcastIP.sin_addr.s_addr)) &&
+                (ntohl(ip.ip_dst.s_addr) < ntohl(minMultiIP.sin_addr.s_addr) || ntohl(ip.ip_dst.s_addr) > ntohl(maxMultiIP.sin_addr.s_addr)))
```

Comment:
Questo codice non e' granche' efficiente: ci sono un sacco do ntohl inutile che possono essere tolti e semplificati con degli and.Esempio     
if ((a & 0xF0000000) == 0xE0000000 /* 224.0.0.0/4 */)
      addr.multicastIP = true;